napari/utils/chunk/_loader.py

```diff
@@ -0,0 +1,480 @@
+Viewer instances. There are two main reasons we do this instead of one
+ChunkLoader per Viewer:
+
+1. We size the ChunkCache as a fraction of RAM, so having more than one
+   cache would use too much RAM.
+
+2. We might size the thread pool for optimal performance, and having
+   multiple pools would result in more workers than we want.
+
+Think of the ChunkLoader as a shared resource like "the filesystem" where
+multiple clients can be access it at the same time, but it is the interface
+to just one physical resource.
+"""
+import logging
+from concurrent import futures
+from contextlib import contextmanager
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+from ...types import ArrayLike
+from ...utils.event import EmitterGroup
+from ..perf import perf_timer
+from ._cache import ChunkCache
+from ._config import async_config
+from ._delay_queue import DelayQueue
+from ._info import LayerInfo, LoadType
+from ._request import ChunkKey, ChunkRequest
+
+LOGGER = logging.getLogger("ChunkLoader")
+
+# Executor for either a thread pool or a process pool.
+PoolExecutor = Union[futures.ThreadPoolExecutor, futures.ProcessPoolExecutor]
+
+
+def _chunk_loader_worker(request: ChunkRequest) -> ChunkRequest:
+    """This is the worker thread or process that loads the array.
+
+    We call np.asarray() in a worker because it might lead to IO or
+    computation which would block the GUI thread.
+
+    Parameters
+    ----------
+    request : ChunkRequest
+        The request to load.
+    """
+    request.load_chunks()  # loads all chunks in the request
+    return request
+
+
+def _create_executor(use_processes: bool, num_workers: int) -> PoolExecutor:
+    """Return the thread or process pool executor.
+
+    Parameters
+    ----------
+    use_processes : bool
+        If True use processes, otherwise threads.
+    num_workers : int
+        The number of worker threads or processes.
+    """
+    if use_processes:
+        return futures.ProcessPoolExecutor(max_workers=num_workers)
+    return futures.ThreadPoolExecutor(max_workers=num_workers)
+
+
+class ChunkLoader:
+    """Loads chunks in worker threads or processes.
+
+    We cannot call np.asarray() in the GUI thread because it might block on
+    IO or a computation. So the ChunkLoader calls np.asarray() for us in
+    a worker thread or process.
+
+    Attributes
+    ----------
+    synchronous : bool
+        If True LoadType.AUTO layers are loading synchronously in the GUI thread.
+    load_seconds : float
+        Sleep this long in the worker during a load.
+    executor : Union[ThreadPoolExecutor, ProcessPoolExecutor]
+        Our thread or process pool executor.
+    futures : FutureMap
+        In progress futures for each layer (data_id).
+    cache : ChunkCache
+        Cache of previously loaded chunks.
+    events : EmitterGroup
+        We only signal one event: chunk_loaded.
+    """
+
+    FutureMap = Dict[int, List[futures.Future]]
```

Comment:
The problem with that is we have to write `Dict[int, List[futures.Future]]` in two places, in the code and that repeat that mouthful in the doc string.

If we write `FutureMap` in both places you you can see that it's the same thing. Instead of looking at `Dict[int, List[futures.Future]]` in those places and having to check letter by letter to see if they are the same or just very similar.

If the docstring was generated from the code then I'd totally agree. Then it would be written in one place only.  I can certainly change it if it's a big deal, but it seems better the way it is to me and my eye.

Reply:
> The problem with that is we have to write Dict[int, List[futures.Future]] in two places, in the code and that repeat that mouthful in the doc string.

why do you think that?  if you're thinking that you need to annotate it inside of `__init__` again, then you're incorrect...  you only need it once:

```python
class ChunkLoader:
    layer_map: Dict[int, LayerInfo]

    def __init__(self) -> None:
        self.layer_map = {}
        reveal_type(self.layer_map[1])
```


<img width="463" alt="Screen Shot 2020-08-17 at 4 49 09 PM" src="https://user-images.githubusercontent.com/1609449/90443074-a3c65380-e0a9-11ea-85fd-a652a7111355.png">

>  I can certainly change it if it's a big deal, but it seems better the way it is to me and my eye.

I'll go with the PEP here.  please change

Reply:
if you _did_ however need to use it more than once, then I absolutely fine using an alias... just don't assign it as a class attribute.  assign it as a module level variable and use it inside of the class

Reply:
It's used twice in the code already. When I search for the string `FutureMap` and I get three hits. The first where it is defined and then the two places it's used, one in a docstring and one in the code:

<img width="395" alt="Screen Shot 2020-08-17 at 6 01 01 PM" src="https://user-images.githubusercontent.com/4163446/90448629-c1001f80-e0b3-11ea-85fe-0abba464aa50.png">

Are you looking at a different version of the code? I'm looking at the `feature/async-image-loading` branch.

Reply:
Also your last example shows `layer_map` as a class attribute. Is that the language you are suggesting, that it should be class attribute instead of an instance attribute as it is today? That seems totally wrong. While we only expect to have one `ChunkLoader` the `layer_map` is still a property of that instance of the loader, that's why it's an instance attribute today. What is your reasoning to want to change it to a class attributed? I've lost the plot here.

Reply:
That is, `layer_map` is an instance attribute right now. I don't see why we'd change it to a class attribute.

Reply:
I'll leave `layer_map` as in instance variable for now. We can always revisit it.